client/karma.js

```diff
@@ -66,8 +66,24 @@ function Karma (socket, iframe, opener, navigator, location) {
       } else if (url !== 'about:blank') {
         var loadScript = function (idx) {
           if (idx < window.__karma__.scriptUrls.length) {
-            var ele = document.createElement('script')
-            ele.src = window.__karma__.scriptUrls[idx]
+            var parser = new DOMParser()
+            // Browsers don't like character <> in string when assigning
```

Comment:
Rather than "inanimate object does not like" can we say eg
// Escape characters with special roles in HTML that appear in the body of tags.

But I don't understand this. How can a DOM parser not support angle bracket parsing?

Reply:
It is actually not the DOM parser issue. Let's say we have only one file (bar/foo.js). We then construct this element that would like to add to the dom (<script type=\"text/javascript\" src=\"/bar/foo.js\" ></script>). When we embed this string into %SCRIPT_URL_ARRAY% in client_with_context.html. The html file will have a line look like this: window.__karma__.scriptUrls = ["<script type=\"text/javascript\" src=\"/bar/foo.js\" ></script>"]. When the browser parse the html file, they stop parsing once it reaches the character "<". They parse it and expect it as an open tag even though it is wrapped with double quotation. Therefore I turn them to hex code which looks like this in the html: window.__karma__.scriptUrls = ["\\x3Cscript type=\"text/javascript\" src=\"/bar/foo.js\" \\x3E\\x3C/script\\x3E"]

Reply:
Wouldn't it be clearer to have the template be an array of objects:
`[{type: "text/javascript", src: "/bar/foo.js" }, {...}]` Then in this code we walk the object and create the script tags.  So no angle brackets in the template means no reason to encode then replace the angle brackets?

Reply:
I also thought about this approach too. However, the script tags html are constructed case by case in the middleware.

`            if (fileType === 'css') {
              scriptTags.push(`<link type="text/css" href="${filePath}" rel="stylesheet">`)
            } else if (fileType === 'dom') {
              scriptTags.push(includedContent.get(file.path))
            } else if (fileType === 'html') {
              scriptTags.push(`<link href="${filePath}" rel="import">`)
            } else {
              const scriptType = (SCRIPT_TYPE[fileType] || 'text/javascript')
              const crossOriginAttribute = includeCrossOriginAttribute ? 'crossorigin="anonymous"' : ''
              scriptTags.push(`<script type="${scriptType}" src="${filePath}" ${crossOriginAttribute}></script>`)
            }
          }`

If this changes in the future, we have to update the logic to construct and reconstruct the script tag object too. I think that actually adds another layer of complexity and maintenance. Therefore, I prefer just escaping and de-escaping  the generated string here (which I think also have closest logic to the regular case).